src/tamp_improv/approaches/improvisational/policies/rl2mpc.py

```diff
@@ -0,0 +1,196 @@
+
+    def initialize(self, env: gym.Env) -> None:
+        """Initialize both RL and MPC components."""
+        self._env = env
+        self.rl_policy.initialize(env)
+        self.mpc_policy.initialize(env)
+
+    def configure_context(self, context: PolicyContext) -> None:
+        """Configure policy context."""
+        self.mpc_policy.configure_context(context)
+
+    def train(self, env: gym.Env, train_data: TrainingData) -> None:
+        """Train RL policy until threshold."""
+        if not hasattr(self, "rl_policy") or not hasattr(self, "mpc_policy"):
+            self.initialize(env)
+
+        # Create callback that will stop training at threshold
+        callback = RL2MPCCallback(
+            reward_threshold=self.config.reward_threshold,
+            window_size=self.config.window_size,
+            check_freq=train_data.config.get("training_record_interval", 100),
+        )
+
+        # Train RL policy
+        print("\nTraining RL component until reward threshold...")
+        self.rl_policy.train(env, train_data, callback=callback)
+
+        # Record if threshold was reached during training
+        # Note: We don't switch to MPC yet - that happens at execution time
+        self._threshold_reached = callback.threshold_reached
+
+    def _initialize_mpc_from_start_state(self) -> None:
+        """Initialize MPC nominal trajectory using RL policy from start
+        state."""
+        if self._env is None or self._start_state is None:
+            raise ValueError("Environment or start state not initialized")
+
+        # Use MPC's wrapped environment for trajectory generation
+        if not hasattr(self.mpc_policy, "env") or self.mpc_policy.env is None:
+            raise ValueError("MPC environment not initialized")
+
+        # Create a separate environment instance for trajectory generation
+        temp_env = deepcopy(self.mpc_policy.env)
+
+        # Generate trajectory from start state using RL
+        current_obs = self._start_state
+        temp_env.reset_from_state(current_obs)  # type: ignore
+        trajectory = []
+
+        for step in range(self.mpc_policy.config.horizon):
+            action = self.rl_policy.get_action(current_obs)
+            trajectory.append(action)
+
+            # Step environment to get next observation
+            next_obs, _, terminated, truncated, _ = temp_env.step(action)
+
+            if terminated or truncated:
+                # If episode ended early, use zero action for remaining steps
+                remaining_steps = self.mpc_policy.config.horizon - (step + 1)
+                zero_action = cast(ActType, np.zeros_like(action))
+                for _ in range(remaining_steps):
+                    trajectory.append(zero_action)
+                break
+
+            current_obs = next_obs
+
+        print(f"Generated MPC trajectory from RL: {trajectory}")
+        trajectory_arr = np.array(trajectory)
+
+        # Set MPC's last solution based on action space type
```

Comment:
this logic is repeated I think -- can you refactor so that this code is only inside MPC policy?